File: src/field.ts
/**
 * Field geometry for RCJA Soccer 2026.
 *
 * Every value here is taken from the RCJA Soccer Rules 2026 (v26.0, 17 Feb 2026)
 * field diagram in section 2 and the prose of sections 2.1-2.4. Units are
 * millimetres throughout; the renderer scales to metres at the last moment.
 *
 * Coordinate system (right-handed, y up):
 *   +x  towards the yellow goal      -x  towards the cyan goal
 *   +z  towards one sideline         -z  towards the other
 *    y  height above the carpet
 * The origin is the centre of the playing area.
 *
 * Rule 2.1.5: all field dimensions carry a 5% tolerance. Robots should be
 * designed for it, so the simulator exposes it rather than pretending the
 * field is exact.
 */

/** Outer extent, wall to wall, including the out area. Diagram: 2430 x 1820. */
export const OUTER_LENGTH = 2430;
export const OUTER_WIDTH = 1820;

/** Rule 2.1.1: white lines are 50 mm thick, set 250 mm in from the walls. */
export const OUT_BAND = 250;
export const LINE_THICKNESS = 50;

/**
 * The playing area is inside the white lines, exclusive of them (rule 2.1.1),
 * so each side loses the 250 mm band plus the 50 mm line.
 * 2430 - 2*300 = 1830 long; 1820 - 2*300 = 1220 wide. The diagram's 915 mm
 * half-length dimension confirms the 1830.
 */
export const PLAY_LENGTH = OUTER_LENGTH - 2 * (OUT_BAND + LINE_THICKNESS); // 1830
export const PLAY_WIDTH = OUTER_WIDTH - 2 * (OUT_BAND + LINE_THICKNESS); // 1220

export const HALF_LENGTH = PLAY_LENGTH / 2; // 915
export const HALF_WIDTH = PLAY_WIDTH / 2; // 610

/** Rule 2.3: goal is 450 mm wide internally, 74 mm deep, crossbar at 140 mm. */
export const GOAL_WIDTH = 450;
/**
 * How thick the goal's own walls are. The rules dimension the goal internally
 * and say nothing about the panels, so this matches the perimeter walls. It is
 * the difference between the 450 mm opening a ball has to fit through and the
 * outside of the structure a robot bumps into, which is 40 mm wider.
 */
export const GOAL_WALL_THICKNESS = 20;
export const GOAL_DEPTH = 74;
export const CROSSBAR_HEIGHT = 140;
export const CROSSBAR_DEPTH = 20;

/** Penalty box from the field diagram: 300 mm deep, 900 mm wide. */
export const PENALTY_DEPTH = 300;
export const PENALTY_WIDTH = 900;

/** Rule 2.1.1: penalty boxes and neutral points are marked in 25 mm black line. */
export const MARKING_THICKNESS = 25;

/**
 * Rule 2.4: two neutral points, plus a centre point used for kick-offs and
 * for repeated lack of progress. The diagram puts the pair on the halfway
 * line, 300 mm either side of centre.
 */
export const NEUTRAL_OFFSET = 300;

export interface Point {
  x: number;
  z: number;
}

/** Rule 2.4.2: the centre point is last in the list; it is the fallback. */
export const NEUTRAL_POINTS: readonly Point[] = [
  { x: 0, z: -NEUTRAL_OFFSET },
  { x: 0, z: NEUTRAL_OFFSET },
  { x: 0, z: 0 },
];

/** Rule 2.2: wall height differs by league; see leagues.ts for the per-league value. */
export const WALL_HEIGHT_NATIONALS = 220;

/** Rule 2.1.2: the out area may be inclined by raising its outer edge 10 mm. */
export const INCLINE_RISE = 10;

/** Rule 2.1.5. */
export const DIMENSION_TOLERANCE = 0.05;

export type GoalSide = 'cyan' | 'yellow';

/** Centre of the mouth of a goal, on the goal line. */
export function goalMouth(side: GoalSide): Point {
  return { x: side === 'cyan' ? -HALF_LENGTH : HALF_LENGTH, z: 0 };
}

/** True if a point lies inside the penalty box defending the given goal. */
export function inPenaltyBox(p: Point, side: GoalSide): boolean {
  const withinWidth = Math.abs(p.z) <= PENALTY_WIDTH / 2;
  if (!withinWidth) return false;
  return side === 'cyan'
    ? p.x <= -HALF_LENGTH + PENALTY_DEPTH
    : p.x >= HALF_LENGTH - PENALTY_DEPTH;
}

/**
 * True if a point is fully outside the playing area, i.e. in the out band.
 * Used for rule 5.7.1.6 (Lightweight/Open: a robot wholly in the out area is
 * damaged) and for the Lightweight/Open ball-out-of-play test (5.9.1).
 */
export function isOutOfPlay(p: Point): boolean {
  return Math.abs(p.x) > HALF_LENGTH || Math.abs(p.z) > HALF_WIDTH;
}

/** Rule 5.6.2 / 5.9.2: the ball goes to the nearest neutral point. */
export function nearestNeutralPoint(p: Point, occupied: readonly Point[] = []): Point {
  const free = NEUTRAL_POINTS.filter(
    (np) => !occupied.some((o) => Math.hypot(o.x - np.x, o.z - np.z) < 110),
  );
  const candidates = free.length > 0 ? free : NEUTRAL_POINTS;
  let best = candidates[0]!;
  let bestDist = Infinity;
  for (const np of candidates) {
    const d = Math.hypot(np.x - p.x, np.z - p.z);
    if (d < bestDist) {
      bestDist = d;
      best = np;
    }
  }
  return best;
}

/**
 * Longitudinal placement of the goals.
 *
 * The field diagram in section 2 puts the goal mouth on the INNER edge of the
 * white line - the goal line proper - so the mouth is flush with the edge of
 * the playing area and the goal itself stands in the out area, its 74 mm of
 * depth eating into the 250 mm band between the line and the end wall.
 *
 * That leaves 226 mm of out area behind the goal. Rule 2.3.6 closes it off by
 * carrying the goal's side walls back to the end wall, so the goal and those
 * walls together are one solid block running from the goal line to the wall;
 * `collideWithPerimeter` treats them as exactly that. The out area either side
 * of the block stays open carpet, reachable like the out area at the touchlines.
 */
export const WALL_X = OUTER_LENGTH / 2; // 1215
export const WALL_Z = OUTER_WIDTH / 2; // 910
export const GOAL_MOUTH_X = HALF_LENGTH; // 915
export const GOAL_BACK_X = GOAL_MOUTH_X + GOAL_DEPTH; // 989
export const HALF_GOAL_WIDTH = GOAL_WIDTH / 2; // 225
/** Half the outside of the goal structure, walls included. */
export const HALF_GOAL_SHELL = HALF_GOAL_WIDTH + GOAL_WALL_THICKNESS; // 245

/**
 * True if a ball centre lies between the posts, i.e. over the mouth opening
 * (rule 2.3: goal is 450 mm wide internally). This is the referee's question
 * "is the ball over the goal", and it deliberately tests the whole opening
 * rather than a ball-radius-shortened one: whether the ball actually gets in
 * and scores is decided separately, by the radius margin in
 * `collideWithPerimeter`, and a ball that clips the mouth edge and rebounds
 * back onto the field must not be called out of play (5.9.1) for coming to
 * rest over the line between the posts.
 */
export function withinGoalMouth(z: number): boolean {
  return Math.abs(z) <= HALF_GOAL_WIDTH;
}

File: python/rcja_soccer/field.py
"""The field, from the rulebook.

Nothing here is privileged information. Every number is printed in the RCJA
Soccer Rules 2026 field diagram and section 2, a team measures them off the
real table before a competition, and a robot that knows where the goal is is
not cheating — it is a robot whose programmer read the rules.

What a robot still does not get is where it *is*, where the ball is, or where
anybody else is. Those have to come out of the sensors. This module only says
what the world is shaped like.

Coordinates match the simulator's own: +x towards the yellow goal, +z towards
one sideline, origin at the centre of the playing area, millimetres throughout.
"""

from __future__ import annotations

import math

#: Rule 2.1.1. The playing area is inside the white lines, exclusive of them.
HALF_LENGTH = 915.0
HALF_WIDTH = 610.0

#: The 50 mm white line sits OUTSIDE the playing area, between HALF_LENGTH and
#: HALF_LENGTH + LINE_THICKNESS. Crossing it is not in itself an offence: rule
#: 5.7.1.6 only punishes a robot that is *wholly* in the out area, and there is
#: a full 250 mm band of out area beyond the line to be wholly inside, at the
#: ends of the field as well as at the touchlines.
LINE_THICKNESS = 50.0
OUT_BAND = 250.0

#: Wall to wall, rule 2.1 diagram: the field is 2430 x 1820 over all.
WALL_X = 1215.0
WALL_Z = 910.0

#: Rule 2.3. The mouth is on the goal line - the inner edge of the white line,
#: flush with the playing area - and the goal runs 74 mm outward from there to
#: the back wall a goal is scored against (rule 5.5.1). Rule 2.3.6 carries the
#: goal's side walls on from there to the end wall, so nothing can pass behind
#: it: driving at your own goal, the thing that stops you is GOAL_MOUTH_X.
GOAL_WIDTH = 450.0
HALF_GOAL_WIDTH = 225.0
GOAL_MOUTH_X = 915.0
GOAL_BACK_X = 989.0

#: From the field diagram.
PENALTY_DEPTH = 300.0
PENALTY_WIDTH = 900.0
NEUTRAL_OFFSET = 300.0
NEUTRAL_POINTS = ((0.0, -NEUTRAL_OFFSET), (0.0, NEUTRAL_OFFSET), (0.0, 0.0))

#: Rule 4.1.2's 220 mm cylinder, and the Open league's 42 mm ball.
ROBOT_RADIUS = 110.0
BALL_RADIUS = 21.0
#: Centre-to-centre distance at which robot and ball are touching.
CONTACT_RANGE = ROBOT_RADIUS + BALL_RADIUS

#: Drivetrain geometry, matching drive.py and the simulator's robot model.
WHEEL_RADIUS = 25.0
MOUNT_RADIUS = 90.0


def attack_x(direction: float) -> float:
    """The x of the goal line this attack direction is shooting at.

    Takes ``s.attack_direction`` (+1 or -1), not a team colour: rule 1.4/5.4
    swaps ends at half-time, so which goal a team shoots at is not fixed for
    the whole match the way its colour is.
    """
    return HALF_LENGTH if direction > 0 else -HALF_LENGTH


def defend_x(direction: float) -> float:
    """The x of the goal line this attack direction is defending."""
    return -attack_x(direction)


def attack_heading(direction: float) -> float:
    """Field heading pointing up the field, towards the opponent's goal."""
    return 0.0 if direction > 0 else math.pi


def their_goal(direction: float) -> str:
    """The camera's name for the goal this attack direction is shooting at."""
    return "yellow" if direction > 0 else "cyan"


def our_goal(direction: float) -> str:
    """The camera's name for the goal this attack direction is defending."""
    return "cyan" if direction > 0 else "yellow"


def other_side(team: str) -> str:
    """The other team's id.

    Team ids are the robot colours — violet or lime; the goals are *places*
    whose paint lives at a fixed place for the whole match, so which name a
    robot calls the goal it is shooting at is a field fact found from the
    frame, not a team fact.
    """
    return "lime" if team == "violet" else "violet"


def goal_centre(side: str) -> tuple[float, float]:
    """Centre of a goal mouth, on the goal line — what the camera ranges to."""
    return (-HALF_LENGTH if side == "cyan" else HALF_LENGTH, 0.0)


def in_penalty_box(x: float, z: float, side: str, slack: float = 0.0) -> bool:
    """Whether a point is inside the penalty box defending the given goal.

    ``side`` names a goal by its paint, which lives at a fixed place; pass the
    frame's ``my_colour`` to get this team's own box in either half.
    """
    if abs(z) > PENALTY_WIDTH / 2 + slack:
        return False
    if side == "cyan":
        return x <= -HALF_LENGTH + PENALTY_DEPTH + slack
    return x >= HALF_LENGTH - PENALTY_DEPTH - slack


def is_out(x: float, z: float) -> bool:
    """Whether a point is outside the playing area, rule 5.9.1's test for Open."""
    return abs(x) > HALF_LENGTH or abs(z) > HALF_WIDTH


def shot_range(bx: float, bz: float, heading: float, side: str) -> float | None:
    """How far a kick from (bx, bz) along a field heading would travel to go in.

    None if it would not go in at all. The kicker fires the ball along the
    robot's *heading*, not along the line from the ball to wherever the program
    was aiming, so this is the question that actually decides whether a shot is
    a shot: ray-cast from the ball and see which boundary it reaches first.

    The distance matters as much as the answer. A ball that leaves the playing
    area goes to a neutral point (rule 5.9.2), and a shot from the far side of
    the halfway line has a keeper, an opposing striker and two metres of carpet
    to survive. Firing one is not ambition, it is a gift.
    """
    dx = math.cos(heading)
    dz = math.sin(heading)
    goal_x, _ = goal_centre(side)

    if dx * goal_x <= 0:
        return None  # pointed away from that goal entirely
    t_goal = (goal_x - bx) / dx
    if t_goal <= 0:
        return None

    # A sideline reached first ends the shot there, out of play.
    if abs(dz) > 1e-6:
        t_side = (math.copysign(HALF_WIDTH, dz) - bz) / dz
        if 0 < t_side < t_goal:
            return None

    # Rule 5.5.1 wants the back wall of the goal; the mouth is the opening it
    # has to fit through, and the ball needs its own width of room.
    if abs(bz + dz * t_goal) > HALF_GOAL_WIDTH - BALL_RADIUS:
        return None
    return t_goal


def kick_lands_in_goal(bx: float, bz: float, heading: float, side: str) -> bool:
    """Whether a kick from here along this heading goes in, at any range."""
    return shot_range(bx, bz, heading, side) is not None


def pass_lands_near(
    bx: float, bz: float, heading: float, target_x: float, target_z: float, tolerance: float = 140.0
) -> float | None:
    """How far a kick from (bx, bz) along heading travels before passing near a teammate.

    The same ray-cast `shot_range` uses, aimed at a robot instead of a goal
    mouth: the kicker fires along the heading, not along the line to wherever
    the program was aiming, so whether a pass finds a teammate is a question
    about the heading, not about where the teammate happens to be standing.
    None if the line never comes within `tolerance` of them, runs backwards,
    or leaves the playing area before it gets there.
    """
    dx = math.cos(heading)
    dz = math.sin(heading)
    t = (target_x - bx) * dx + (target_z - bz) * dz
    if t <= 0:
        return None
    miss = math.hypot(bx + dx * t - target_x, bz + dz * t - target_z)
    if miss > tolerance:
        return None
    if is_out(bx + dx * t, bz + dz * t):
        return None
    return t


def clear_is_safe(bx: float, bz: float, heading: float) -> bool:
    """Whether a clearance along this heading stays in play for a while.

    A keeper's clearance does not have to end up anywhere in particular, but it
    must not simply put the ball out over the nearest sideline, which is the
    usual way a keeper gives possession straight back.
    """
    dx = math.cos(heading)
    dz = math.sin(heading)
    reach = 700.0
    return not is_out(bx + dx * reach, bz + dz * reach)
